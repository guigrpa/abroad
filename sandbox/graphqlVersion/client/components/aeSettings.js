// @flow

/* eslint-disable react/no-access-state-in-setstate */

import timm from 'timm';
import React from 'react';
import Relay, { graphql } from 'react-relay';
import pick from 'lodash/pick';
import {
  flexContainer,
  Icon,
  Select,
  Checkbox,
  TextInput,
  Modal,
  notify,
} from 'giu';
import type { ConfigT } from '../../common/types';
import _t from '../../translate';
import updateConfig from '../mutations/updateConfig';
import { LANG_OPTIONS } from '../gral/constants';
import { mutate } from './helpers';

const STATE_ATTRS = [
  'langs',
  'srcPaths',
  'srcExtensions',
  'msgFunctionNames',
  'msgRegexps',
  'fJsOutput',
];

// ==========================================
// Component declarations
// ==========================================
type Props = {
  lang: string,
  onChangeLang: (str: string) => void,
  onClose: () => void,
  // Relay
  relay: Object,
  config: ConfigT,
};

const gqlFragments = graphql`
  fragment aeSettings_config on Config {
    langs
    srcPaths
    srcExtensions
    msgFunctionNames
    msgRegexps
    fMinify
    fJsOutput
    fJsonOutput
    fReactIntlOutput
  }
`;

type FlexDirection = 'row' | 'column';

// ==========================================
// Component
// ==========================================
class Settings extends React.Component {
  props: Props;
  state: {
    langs: Array<string>,
    srcPaths: Array<string>,
    srcExtensions: Array<string>,
    msgFunctionNames: Array<string>,
    msgRegexps: Array<string>,
    fJsOutput: boolean,
  };
  refLang: ?Object;
  refMinify: ?Object;
  refReactIntlOutput: ?Object;
  refJsonOutput: ?Object;

  constructor(props: Props) {
    super(props);
    // - For arrays without IDs, it's better if we keep the current state at this level,
    // rather than relying on `giu`.
    // - For `fJsOutput`, we need to track it since more than one input depend on it.
    // - For other attributes (`lang`, `fMinify`...), we can
    // leave state handling entirely to `giu`, and fetch the value when the user clicks on
    // Save.
    this.state = pick(props.config, STATE_ATTRS);
  }

  // ------------------------------------------
  render() {
    const buttons = [
      { label: _t('button_Cancel'), onClick: this.onCancel, left: true },
      { label: _t('button_Save'), onClick: this.onSave, defaultButton: true },
    ];
    return (
      <Modal
        buttons={buttons}
        onEsc={this.onCancel}
        onClickBackdrop={this.onCancel}
      >
        {this.renderConfig()}
      </Modal>
    );
  }

  renderConfig() {
    const { lang } = this.props;
    const { fMinify, fJsonOutput, fReactIntlOutput } = this.props.config;
    const { fJsOutput } = this.state;
    return (
      <div>
        <div style={style.configLine}>
          <label htmlFor="lang">{_t('settingsForm_Mady language:')}</label>{' '}
          <Select
            ref={c => {
              this.refLang = c;
            }}
            value={lang}
            items={LANG_OPTIONS}
            required
          />
        </div>
        <div style={style.listLabel}>
          {_t('settingsForm_Languages (BCP47 codes):')}
        </div>
        {this.renderList({
          id: 'langs',
          dir: 'row',
          Component: TextInput,
          placeholder: 'e.g. es-ES, caâ€¦',
          width: 80,
        })}
        <div style={style.listLabel}>{_t('settingsForm_Source paths:')}</div>
        {this.renderList({
          id: 'srcPaths',
          dir: 'column',
          Component: TextInput,
          placeholder: 'e.g. src/client',
          width: 300,
        })}
        <div style={style.listLabel}>
          {_t('settingsForm_Source extensions:')}
        </div>
        {this.renderList({
          id: 'srcExtensions',
          dir: 'row',
          Component: TextInput,
          placeholder: 'e.g. .js',
          width: 60,
        })}
        <div style={style.listLabel}>
          {_t('settingsForm_Message translation functions to look for:')}
        </div>
        {this.renderList({
          id: 'msgFunctionNames',
          dir: 'row',
          Component: TextInput,
          placeholder: 'e.g. _t',
          width: 60,
        })}
        <div style={style.listLabel}>
          {_t(
            'settingsForm_ADVANCED: Additional regular expressions for message parsing:'
          )}{' '}
          <Icon
            icon="info-circle"
            title={_t(
              'settingsForm_Make sure your regular expression has exactly one capture group, for example: (.*?)'
            )}
            style={style.info}
          />
        </div>
        {this.renderList({
          id: 'msgRegexps',
          dir: 'column',
          Component: TextInput,
          placeholder: 'e.g. {{\\s*(.*?)\\s*}}',
          width: 300,
        })}
        <div style={style.configLine}>
          <div>{_t('settingsForm_Output:')}</div>
          <div style={style.indented}>
            <Checkbox
              id="fJsOutput"
              value={fJsOutput}
              onChange={(ev, val) => this.setState({ fJsOutput: val })}
            />
            <label htmlFor="fJsOutput">
              {_t(
                "settingsForm_JavaScript module (required if you use Mady's translation function)"
              )}
            </label>{' '}
            <Checkbox
              ref={c => {
                this.refMinify = c;
              }}
              id="fMinify"
              disabled={!fJsOutput}
              value={fMinify}
            />
            <label htmlFor="fMinify">{_t('settingsForm_Minified')}</label>
          </div>
          <div style={style.indented}>
            <Checkbox
              ref={c => {
                this.refReactIntlOutput = c;
              }}
              id="fReactIntlOutput"
              value={fReactIntlOutput}
            />
            <label htmlFor="fReactIntlOutput">
              {_t('settingsForm_React Intl JSON file')}
            </label>
          </div>
          <div style={style.indented}>
            <Checkbox
              ref={c => {
                this.refJsonOutput = c;
              }}
              id="fJsonOutput"
              value={fJsonOutput}
            />
            <label htmlFor="fJsonOutput">
              {_t('settingsForm_Generic JSON file')}
            </label>
          </div>
        </div>
      </div>
    );
  }

  /* eslint-disable react/no-unused-prop-types, react/no-array-index-key */
  renderList({
    id,
    dir,
    Component,
    placeholder,
    width,
  }: {
    id: string,
    dir: FlexDirection,
    Component: any,
    placeholder: string,
    width: number,
  }) {
    const values = this.state[id];
    return (
      <div style={style.list(dir)}>
        {values.map((value, idx) => (
          <div key={idx} style={style.listItem(dir)}>
            <Component
              id={`${id}.${idx}`}
              value={value}
              placeholder={placeholder}
              onChange={this.onUpdateListItem}
              required
              errorZ={52}
              style={style.input(width)}
            />
            &nbsp;
            <Icon
              id={`${id}.${idx}`}
              icon="remove"
              onClick={this.onRemoveListItem}
              style={style.remove}
            />
          </div>
        ))}
        <Icon
          id={id}
          icon="plus"
          onClick={this.onCreateListItem}
          style={style.add}
        />
      </div>
    );
  }
  /* eslint-enable react/no-unused-prop-types,react/no-array-index-key */

  // ------------------------------------------
  onCreateListItem = (ev: SyntheticEvent) => {
    if (!(ev.currentTarget instanceof HTMLElement)) return;
    const { id } = ev.currentTarget;
    const newList = timm.addLast(this.state[id], '');
    this.setState({ [id]: newList });
  };

  onRemoveListItem = (ev: SyntheticEvent) => {
    if (!(ev.currentTarget instanceof HTMLElement)) return;
    const [id, idx] = ev.currentTarget.id.split('.');
    const newList = timm.removeAt(this.state[id], Number(idx));
    this.setState({ [id]: newList });
  };

  onUpdateListItem = (ev: SyntheticEvent) => {
    if (!(ev.currentTarget instanceof HTMLInputElement)) return;
    const value = ev.currentTarget.value;
    const [id, idx] = ev.currentTarget.id.split('.');
    const newList = timm.replaceAt(this.state[id], Number(idx), value);
    this.setState({ [id]: newList });
  };

  onCancel = () => {
    this.props.onClose();
  };
  onSave = () => {
    // Save lang
    if (this.refLang == null) return;
    const lang = this.refLang.getValue();
    if (lang !== this.props.lang) this.props.onChangeLang(lang);

    // Save other settings
    const attrs = pick(this.state, STATE_ATTRS);
    if (this.refMinify == null) return;
    attrs.fMinify = this.refMinify.getValue();
    if (this.refReactIntlOutput == null) return;
    attrs.fReactIntlOutput = this.refReactIntlOutput.getValue();
    if (this.refJsonOutput == null) return;
    attrs.fJsonOutput = this.refJsonOutput.getValue();
    mutate({
      description: 'Click on Save settings',
      environment: this.props.relay.environment,
      mutationOptions: updateConfig({
        config: this.props.config,
        attrs,
      }),
      onFailure: () => {
        notify({
          msg: _t('error_Configuration could not be saved'),
          type: 'error',
          icon: 'save',
        });
      },
    });
    this.props.onClose();
  };
}

// ------------------------------------------
const style = {
  listLabel: {
    marginTop: 7,
    marginBottom: 3,
  },
  list: (dir: FlexDirection) =>
    flexContainer(dir, {
      marginLeft: 15,
    }),
  listItem: (dir: FlexDirection) => ({
    padding: '0px 2px',
    marginTop: dir === 'column' ? 1 : undefined,
    marginRight: 10,
    whiteSpace: 'nowrap',
  }),
  input: width => ({ width }),
  add: {
    display: 'inline-block',
    marginTop: 5,
    color: '#444',
  },
  remove: { color: '#444' },
  configLine: {
    marginTop: 7,
  },
  indented: {
    marginLeft: 15,
  },
  info: {
    cursor: 'pointer',
  },
};

// ==========================================
// Public API
// ==========================================
const Container = Relay.createFragmentContainer(Settings, gqlFragments);
export default Container;
export { Settings as _Settings };
