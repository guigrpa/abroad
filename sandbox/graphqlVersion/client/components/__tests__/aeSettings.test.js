/* eslint-env jest */
import React from 'react';
import renderer from 'react-test-renderer';
import { Floats } from 'giu';
import { _Settings as Settings } from '../aeSettings';

// https://github.com/facebook/react/issues/7386#issuecomment-238091398
jest.mock('react-dom');

describe('Settings', () => {
  it('01 normal', () => {
    const config = {
      langs: ['en', 'es'],
      srcPaths: ['src'],
      srcExtensions: ['.js', '.jsx'],
      msgFunctionNames: ['_t'],
      msgRegexps: [],
      fMinify: false,
    };
    const tree = renderer
      .create(
        <div>
          <Floats />
          <Settings
            lang="en"
            config={config}
            onChangeLang={() => {}}
            onClose={() => {}}
          />
        </div>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
