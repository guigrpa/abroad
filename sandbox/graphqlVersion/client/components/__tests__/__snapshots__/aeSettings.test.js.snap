// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Settings 01 normal 1`] = `
<div>
  <div
    className="giu-floats"
    style={
      Object {
        "height": 0,
        "left": 0,
        "position": "absolute",
        "top": 0,
        "width": 0,
      }
    }
  />
  <div
    className="giu-modal"
    onClick={[Function]}
    onKeyDown={[Function]}
    style={
      Object {
        "height": "100vh",
        "left": 0,
        "position": "fixed",
        "top": 0,
        "width": "100vw",
        "zIndex": 50,
      }
    }
  >
    <div
      className="giu-backdrop"
      onClick={[Function]}
      onTouchMove={[Function]}
      onWheel={[Function]}
      style={
        Object {
          "backgroundColor": "white",
          "height": "100vh",
          "left": 0,
          "position": "fixed",
          "top": 0,
          "width": "100vw",
        }
      }
    />
    <div
      style={
        Object {
          "display": "flex",
          "flexDirection": "row",
          "left": 0,
          "position": "fixed",
          "top": "5vh",
          "width": "100%",
        }
      }
    >
      <div
        onClick={[Function]}
        onTouchMove={[Function]}
        onWheel={[Function]}
        style={
          Object {
            "WebkitFlex": 1,
            "flex": 1,
          }
        }
      />
      <div
        onScroll={[Function]}
        onWheel={[Function]}
        style={
          Object {
            "backgroundColor": "white",
            "borderRadius": 2,
            "boxShadow": "0 3px 6px rgba(0,0,0,0.16), 0 3px 6px rgba(0,0,0,0.23)",
            "maxHeight": "90vh",
            "overflowY": "auto",
            "padding": 20,
          }
        }
      >
        <input
          autoFocus={true}
          style={
            Object {
              "cursor": "default",
              "height": 0,
              "left": 8,
              "opacity": 0,
              "padding": 0,
              "pointerEvents": "none",
              "position": "fixed",
              "top": 8,
              "width": 0,
              "zIndex": -80,
            }
          }
        />
        <div>
          <div
            style={
              Object {
                "marginTop": 7,
              }
            }
          >
            <label
              htmlFor="lang"
            >
              Mady language:
            </label>
             
            <select
              className="giu-select-native"
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              style={
                Object {
                  "backgroundColor": "default",
                  "border": "default",
                  "color": "inherit",
                  "fontFamily": "inherit",
                  "fontSize": "inherit",
                  "fontWeight": "inherit",
                }
              }
              type="native"
              value="\\"en\\""
            >
              <option
                id="\\"en\\""
                value="\\"en\\""
              >
                English
              </option>
              <option
                id="\\"es\\""
                value="\\"es\\""
              >
                Español
              </option>
              <option
                id="\\"ca\\""
                value="\\"ca\\""
              >
                Català
              </option>
            </select>
          </div>
          <div
            style={
              Object {
                "marginBottom": 3,
                "marginTop": 7,
              }
            }
          >
            Languages (BCP47 codes):
          </div>
          <div
            style={
              Object {
                "display": "flex",
                "flexDirection": "row",
                "marginLeft": 15,
              }
            }
          >
            <div
              style={
                Object {
                  "marginRight": 10,
                  "marginTop": undefined,
                  "padding": "0px 2px",
                  "whiteSpace": "nowrap",
                }
              }
            >
              <input
                className="giu-text-input"
                id="langs.0"
                onBlur={[Function]}
                onChange={[Function]}
                onFocus={[Function]}
                placeholder="e.g. es-ES, ca…"
                style={
                  Object {
                    "backgroundColor": "white",
                    "border": "1px solid #999999",
                    "color": "inherit",
                    "fontFamily": "inherit",
                    "fontSize": "inherit",
                    "fontWeight": "inherit",
                    "width": 80,
                  }
                }
                type="text"
                value="en"
              />
               
              <i
                className="giu-icon fa fa-remove"
                id="langs.0"
                onClick={[Function]}
                style={
                  Object {
                    "color": "#444",
                    "cursor": "pointer",
                    "fontSize": undefined,
                    "letterSpacing": "normal",
                  }
                }
              />
            </div>
            <div
              style={
                Object {
                  "marginRight": 10,
                  "marginTop": undefined,
                  "padding": "0px 2px",
                  "whiteSpace": "nowrap",
                }
              }
            >
              <input
                className="giu-text-input"
                id="langs.1"
                onBlur={[Function]}
                onChange={[Function]}
                onFocus={[Function]}
                placeholder="e.g. es-ES, ca…"
                style={
                  Object {
                    "backgroundColor": "white",
                    "border": "1px solid #999999",
                    "color": "inherit",
                    "fontFamily": "inherit",
                    "fontSize": "inherit",
                    "fontWeight": "inherit",
                    "width": 80,
                  }
                }
                type="text"
                value="es"
              />
               
              <i
                className="giu-icon fa fa-remove"
                id="langs.1"
                onClick={[Function]}
                style={
                  Object {
                    "color": "#444",
                    "cursor": "pointer",
                    "fontSize": undefined,
                    "letterSpacing": "normal",
                  }
                }
              />
            </div>
            <i
              className="giu-icon fa fa-plus"
              id="langs"
              onClick={[Function]}
              style={
                Object {
                  "color": "#444",
                  "cursor": "pointer",
                  "display": "inline-block",
                  "fontSize": undefined,
                  "letterSpacing": "normal",
                  "marginTop": 5,
                }
              }
            />
          </div>
          <div
            style={
              Object {
                "marginBottom": 3,
                "marginTop": 7,
              }
            }
          >
            Source paths:
          </div>
          <div
            style={
              Object {
                "display": "flex",
                "flexDirection": "column",
                "marginLeft": 15,
              }
            }
          >
            <div
              style={
                Object {
                  "marginRight": 10,
                  "marginTop": 1,
                  "padding": "0px 2px",
                  "whiteSpace": "nowrap",
                }
              }
            >
              <input
                className="giu-text-input"
                id="srcPaths.0"
                onBlur={[Function]}
                onChange={[Function]}
                onFocus={[Function]}
                placeholder="e.g. src/client"
                style={
                  Object {
                    "backgroundColor": "white",
                    "border": "1px solid #999999",
                    "color": "inherit",
                    "fontFamily": "inherit",
                    "fontSize": "inherit",
                    "fontWeight": "inherit",
                    "width": 300,
                  }
                }
                type="text"
                value="src"
              />
               
              <i
                className="giu-icon fa fa-remove"
                id="srcPaths.0"
                onClick={[Function]}
                style={
                  Object {
                    "color": "#444",
                    "cursor": "pointer",
                    "fontSize": undefined,
                    "letterSpacing": "normal",
                  }
                }
              />
            </div>
            <i
              className="giu-icon fa fa-plus"
              id="srcPaths"
              onClick={[Function]}
              style={
                Object {
                  "color": "#444",
                  "cursor": "pointer",
                  "display": "inline-block",
                  "fontSize": undefined,
                  "letterSpacing": "normal",
                  "marginTop": 5,
                }
              }
            />
          </div>
          <div
            style={
              Object {
                "marginBottom": 3,
                "marginTop": 7,
              }
            }
          >
            Source extensions:
          </div>
          <div
            style={
              Object {
                "display": "flex",
                "flexDirection": "row",
                "marginLeft": 15,
              }
            }
          >
            <div
              style={
                Object {
                  "marginRight": 10,
                  "marginTop": undefined,
                  "padding": "0px 2px",
                  "whiteSpace": "nowrap",
                }
              }
            >
              <input
                className="giu-text-input"
                id="srcExtensions.0"
                onBlur={[Function]}
                onChange={[Function]}
                onFocus={[Function]}
                placeholder="e.g. .js"
                style={
                  Object {
                    "backgroundColor": "white",
                    "border": "1px solid #999999",
                    "color": "inherit",
                    "fontFamily": "inherit",
                    "fontSize": "inherit",
                    "fontWeight": "inherit",
                    "width": 60,
                  }
                }
                type="text"
                value=".js"
              />
               
              <i
                className="giu-icon fa fa-remove"
                id="srcExtensions.0"
                onClick={[Function]}
                style={
                  Object {
                    "color": "#444",
                    "cursor": "pointer",
                    "fontSize": undefined,
                    "letterSpacing": "normal",
                  }
                }
              />
            </div>
            <div
              style={
                Object {
                  "marginRight": 10,
                  "marginTop": undefined,
                  "padding": "0px 2px",
                  "whiteSpace": "nowrap",
                }
              }
            >
              <input
                className="giu-text-input"
                id="srcExtensions.1"
                onBlur={[Function]}
                onChange={[Function]}
                onFocus={[Function]}
                placeholder="e.g. .js"
                style={
                  Object {
                    "backgroundColor": "white",
                    "border": "1px solid #999999",
                    "color": "inherit",
                    "fontFamily": "inherit",
                    "fontSize": "inherit",
                    "fontWeight": "inherit",
                    "width": 60,
                  }
                }
                type="text"
                value=".jsx"
              />
               
              <i
                className="giu-icon fa fa-remove"
                id="srcExtensions.1"
                onClick={[Function]}
                style={
                  Object {
                    "color": "#444",
                    "cursor": "pointer",
                    "fontSize": undefined,
                    "letterSpacing": "normal",
                  }
                }
              />
            </div>
            <i
              className="giu-icon fa fa-plus"
              id="srcExtensions"
              onClick={[Function]}
              style={
                Object {
                  "color": "#444",
                  "cursor": "pointer",
                  "display": "inline-block",
                  "fontSize": undefined,
                  "letterSpacing": "normal",
                  "marginTop": 5,
                }
              }
            />
          </div>
          <div
            style={
              Object {
                "marginBottom": 3,
                "marginTop": 7,
              }
            }
          >
            Message translation functions to look for:
          </div>
          <div
            style={
              Object {
                "display": "flex",
                "flexDirection": "row",
                "marginLeft": 15,
              }
            }
          >
            <div
              style={
                Object {
                  "marginRight": 10,
                  "marginTop": undefined,
                  "padding": "0px 2px",
                  "whiteSpace": "nowrap",
                }
              }
            >
              <input
                className="giu-text-input"
                id="msgFunctionNames.0"
                onBlur={[Function]}
                onChange={[Function]}
                onFocus={[Function]}
                placeholder="e.g. _t"
                style={
                  Object {
                    "backgroundColor": "white",
                    "border": "1px solid #999999",
                    "color": "inherit",
                    "fontFamily": "inherit",
                    "fontSize": "inherit",
                    "fontWeight": "inherit",
                    "width": 60,
                  }
                }
                type="text"
                value="_t"
              />
               
              <i
                className="giu-icon fa fa-remove"
                id="msgFunctionNames.0"
                onClick={[Function]}
                style={
                  Object {
                    "color": "#444",
                    "cursor": "pointer",
                    "fontSize": undefined,
                    "letterSpacing": "normal",
                  }
                }
              />
            </div>
            <i
              className="giu-icon fa fa-plus"
              id="msgFunctionNames"
              onClick={[Function]}
              style={
                Object {
                  "color": "#444",
                  "cursor": "pointer",
                  "display": "inline-block",
                  "fontSize": undefined,
                  "letterSpacing": "normal",
                  "marginTop": 5,
                }
              }
            />
          </div>
          <div
            style={
              Object {
                "marginBottom": 3,
                "marginTop": 7,
              }
            }
          >
            ADVANCED: Additional regular expressions for message parsing:
             
            <i
              className="giu-icon fa fa-info-circle"
              style={
                Object {
                  "color": undefined,
                  "cursor": "pointer",
                  "fontSize": undefined,
                  "letterSpacing": "normal",
                }
              }
              title="Make sure your regular expression has exactly one capture group, for example: (.*?)"
            />
          </div>
          <div
            style={
              Object {
                "display": "flex",
                "flexDirection": "column",
                "marginLeft": 15,
              }
            }
          >
            <i
              className="giu-icon fa fa-plus"
              id="msgRegexps"
              onClick={[Function]}
              style={
                Object {
                  "color": "#444",
                  "cursor": "pointer",
                  "display": "inline-block",
                  "fontSize": undefined,
                  "letterSpacing": "normal",
                  "marginTop": 5,
                }
              }
            />
          </div>
          <div
            style={
              Object {
                "marginTop": 7,
              }
            }
          >
            <div>
              Output:
            </div>
            <div
              style={
                Object {
                  "marginLeft": 15,
                }
              }
            >
              <input
                checked={false}
                className="giu-checkbox"
                id="fJsOutput"
                onBlur={[Function]}
                onChange={[Function]}
                onFocus={[Function]}
                type="checkbox"
              />
              <label
                htmlFor="fJsOutput"
              >
                JavaScript module (required if you use Mady's translation function)
              </label>
               
              <input
                checked={false}
                className="giu-checkbox"
                disabled={true}
                id="fMinify"
                onBlur={[Function]}
                onChange={[Function]}
                onFocus={[Function]}
                tabIndex={-1}
                type="checkbox"
              />
              <label
                htmlFor="fMinify"
              >
                Minified
              </label>
            </div>
            <div
              style={
                Object {
                  "marginLeft": 15,
                }
              }
            >
              <input
                checked={false}
                className="giu-checkbox"
                id="fReactIntlOutput"
                onBlur={[Function]}
                onChange={[Function]}
                onFocus={[Function]}
                type="checkbox"
              />
              <label
                htmlFor="fReactIntlOutput"
              >
                React Intl JSON file
              </label>
            </div>
            <div
              style={
                Object {
                  "marginLeft": 15,
                }
              }
            >
              <input
                checked={false}
                className="giu-checkbox"
                id="fJsonOutput"
                onBlur={[Function]}
                onChange={[Function]}
                onFocus={[Function]}
                type="checkbox"
              />
              <label
                htmlFor="fJsonOutput"
              >
                Generic JSON file
              </label>
            </div>
          </div>
        </div>
        <div
          style={
            Object {
              "display": "flex",
              "flexDirection": "row",
              "marginTop": 10,
              "paddingTop": 10,
            }
          }
        >
          <span
            className="giu-button"
            onClick={[Function]}
            style={
              Object {
                "WebkitUserSelect": "none",
                "border": "1px solid #999999",
                "borderRadius": 5,
                "cursor": "pointer",
                "display": "inline-block",
                "marginRight": 5,
                "padding": "1px 5px",
                "userSelect": "none",
              }
            }
          >
            Cancel
          </span>
          <div
            style={
              Object {
                "WebkitFlex": 1,
                "flex": 1,
              }
            }
          />
          <span
            className="giu-button"
            onClick={[Function]}
            style={
              Object {
                "WebkitUserSelect": "none",
                "border": "1px solid black",
                "borderRadius": 5,
                "cursor": "pointer",
                "display": "inline-block",
                "marginLeft": 5,
                "padding": "1px 5px",
                "userSelect": "none",
              }
            }
          >
            Save
          </span>
        </div>
      </div>
      <div
        onClick={[Function]}
        onTouchMove={[Function]}
        onWheel={[Function]}
        style={
          Object {
            "WebkitFlex": 1,
            "flex": 1,
          }
        }
      />
    </div>
  </div>
</div>
------------HTML PREVIEW---------------
<div>
  <div
    class="giu-floats"
    style="height:0;left:0;position:absolute;top:0;width:0;"
  >
  </div>
  <div
    class="giu-modal"
    style="height:100vh;left:0;position:fixed;top:0;width:100vw;z-index:50;"
  >
    <div
      class="giu-backdrop"
      style="background-color:white;height:100vh;left:0;position:fixed;top:0;width:100vw;"
    >
    </div>
    <div style="display:flex;flex-direction:row;left:0;position:fixed;top:5vh;width:100%;">
      <div style="-webkit-flex:1;flex:1;"></div>
      <div style="background-color:white;border-radius:2px;box-shadow:0 3px 6px rgba(0,0,0,0.16), 0 3px 6px rgba(0,0,0,0.23);max-height:90vh;overflow-y:auto;padding:20px;">
        <input
          autoFocus
          style="cursor:default;height:0;left:8px;opacity:0;padding:0;pointer-events:none;position:fixed;top:8px;width:0;z-index:-80;"
        >
        <div>
          <div style="margin-top:7px;">
            <label for="lang">
              Mady language:
            </label>
             
            <select
              class="giu-select-native"
              style="background-color:default;border:default;color:inherit;font-family:inherit;font-size:inherit;font-weight:inherit;"
              type="native"
              value="\\"en\\""
            >
              <option
                id="\\"en\\""
                value="\\"en\\""
              >
                English
              </option>
              <option
                id="\\"es\\""
                value="\\"es\\""
              >
                Español
              </option>
              <option
                id="\\"ca\\""
                value="\\"ca\\""
              >
                Català
              </option>
            </select>
          </div>
          <div style="margin-bottom:3px;margin-top:7px;">
            Languages (BCP47 codes):
          </div>
          <div style="display:flex;flex-direction:row;margin-left:15px;">
            <div style="margin-right:10px;padding:0px 2px;white-space:nowrap;">
              <input
                class="giu-text-input"
                id="langs.0"
                placeholder="e.g. es-ES, ca…"
                style="background-color:white;border:1px solid #999999;color:inherit;font-family:inherit;font-size:inherit;font-weight:inherit;width:80px;"
                type="text"
                value="en"
              >
               
              <i
                class="giu-icon fa fa-remove"
                id="langs.0"
                style="color:#444;cursor:pointer;letter-spacing:normal;"
              >
              </i>
            </div>
            <div style="margin-right:10px;padding:0px 2px;white-space:nowrap;">
              <input
                class="giu-text-input"
                id="langs.1"
                placeholder="e.g. es-ES, ca…"
                style="background-color:white;border:1px solid #999999;color:inherit;font-family:inherit;font-size:inherit;font-weight:inherit;width:80px;"
                type="text"
                value="es"
              >
               
              <i
                class="giu-icon fa fa-remove"
                id="langs.1"
                style="color:#444;cursor:pointer;letter-spacing:normal;"
              >
              </i>
            </div>
            <i
              class="giu-icon fa fa-plus"
              id="langs"
              style="color:#444;cursor:pointer;display:inline-block;letter-spacing:normal;margin-top:5px;"
            >
            </i>
          </div>
          <div style="margin-bottom:3px;margin-top:7px;">
            Source paths:
          </div>
          <div style="display:flex;flex-direction:column;margin-left:15px;">
            <div style="margin-right:10px;margin-top:1px;padding:0px 2px;white-space:nowrap;">
              <input
                class="giu-text-input"
                id="srcPaths.0"
                placeholder="e.g. src/client"
                style="background-color:white;border:1px solid #999999;color:inherit;font-family:inherit;font-size:inherit;font-weight:inherit;width:300px;"
                type="text"
                value="src"
              >
               
              <i
                class="giu-icon fa fa-remove"
                id="srcPaths.0"
                style="color:#444;cursor:pointer;letter-spacing:normal;"
              >
              </i>
            </div>
            <i
              class="giu-icon fa fa-plus"
              id="srcPaths"
              style="color:#444;cursor:pointer;display:inline-block;letter-spacing:normal;margin-top:5px;"
            >
            </i>
          </div>
          <div style="margin-bottom:3px;margin-top:7px;">
            Source extensions:
          </div>
          <div style="display:flex;flex-direction:row;margin-left:15px;">
            <div style="margin-right:10px;padding:0px 2px;white-space:nowrap;">
              <input
                class="giu-text-input"
                id="srcExtensions.0"
                placeholder="e.g. .js"
                style="background-color:white;border:1px solid #999999;color:inherit;font-family:inherit;font-size:inherit;font-weight:inherit;width:60px;"
                type="text"
                value=".js"
              >
               
              <i
                class="giu-icon fa fa-remove"
                id="srcExtensions.0"
                style="color:#444;cursor:pointer;letter-spacing:normal;"
              >
              </i>
            </div>
            <div style="margin-right:10px;padding:0px 2px;white-space:nowrap;">
              <input
                class="giu-text-input"
                id="srcExtensions.1"
                placeholder="e.g. .js"
                style="background-color:white;border:1px solid #999999;color:inherit;font-family:inherit;font-size:inherit;font-weight:inherit;width:60px;"
                type="text"
                value=".jsx"
              >
               
              <i
                class="giu-icon fa fa-remove"
                id="srcExtensions.1"
                style="color:#444;cursor:pointer;letter-spacing:normal;"
              >
              </i>
            </div>
            <i
              class="giu-icon fa fa-plus"
              id="srcExtensions"
              style="color:#444;cursor:pointer;display:inline-block;letter-spacing:normal;margin-top:5px;"
            >
            </i>
          </div>
          <div style="margin-bottom:3px;margin-top:7px;">
            Message translation functions to look for:
          </div>
          <div style="display:flex;flex-direction:row;margin-left:15px;">
            <div style="margin-right:10px;padding:0px 2px;white-space:nowrap;">
              <input
                class="giu-text-input"
                id="msgFunctionNames.0"
                placeholder="e.g. _t"
                style="background-color:white;border:1px solid #999999;color:inherit;font-family:inherit;font-size:inherit;font-weight:inherit;width:60px;"
                type="text"
                value="_t"
              >
               
              <i
                class="giu-icon fa fa-remove"
                id="msgFunctionNames.0"
                style="color:#444;cursor:pointer;letter-spacing:normal;"
              >
              </i>
            </div>
            <i
              class="giu-icon fa fa-plus"
              id="msgFunctionNames"
              style="color:#444;cursor:pointer;display:inline-block;letter-spacing:normal;margin-top:5px;"
            >
            </i>
          </div>
          <div style="margin-bottom:3px;margin-top:7px;">
            ADVANCED: Additional regular expressions for message parsing:
             
            <i
              class="giu-icon fa fa-info-circle"
              style="cursor:pointer;letter-spacing:normal;"
              title="Make sure your regular expression has exactly one capture group, for example: (.*?)"
            >
            </i>
          </div>
          <div style="display:flex;flex-direction:column;margin-left:15px;">
            <i
              class="giu-icon fa fa-plus"
              id="msgRegexps"
              style="color:#444;cursor:pointer;display:inline-block;letter-spacing:normal;margin-top:5px;"
            >
            </i>
          </div>
          <div style="margin-top:7px;">
            <div>
              Output:
            </div>
            <div style="margin-left:15px;">
              <input
                class="giu-checkbox"
                id="fJsOutput"
                type="checkbox"
              >
              <label for="fJsOutput">
                JavaScript module (required if you use Mady&#39;s translation function)
              </label>
               
              <input
                class="giu-checkbox"
                disabled
                id="fMinify"
                type="checkbox"
              >
              <label for="fMinify">
                Minified
              </label>
            </div>
            <div style="margin-left:15px;">
              <input
                class="giu-checkbox"
                id="fReactIntlOutput"
                type="checkbox"
              >
              <label for="fReactIntlOutput">
                React Intl JSON file
              </label>
            </div>
            <div style="margin-left:15px;">
              <input
                class="giu-checkbox"
                id="fJsonOutput"
                type="checkbox"
              >
              <label for="fJsonOutput">
                Generic JSON file
              </label>
            </div>
          </div>
        </div>
        <div style="display:flex;flex-direction:row;margin-top:10px;padding-top:10px;">
          <span
            class="giu-button"
            style="-webkit-user-select:none;border:1px solid #999999;border-radius:5px;cursor:pointer;display:inline-block;margin-right:5px;padding:1px 5px;user-select:none;"
          >
            Cancel
          </span>
          <div style="-webkit-flex:1;flex:1;"></div>
          <span
            class="giu-button"
            style="-webkit-user-select:none;border:1px solid black;border-radius:5px;cursor:pointer;display:inline-block;margin-left:5px;padding:1px 5px;user-select:none;"
          >
            Save
          </span>
        </div>
      </div>
      <div style="-webkit-flex:1;flex:1;"></div>
    </div>
  </div>
</div>
`;
